File: PSQt/tags/V00-00-13/src/GUILogger.cpp
//--------------------------

#include "PSQt/GUILogger.h"
#include "PSQt/Logger.h"

#include <iostream>  // std::cout
#include <sstream>   // std::stringstream

#include <QIntValidator>

//using namespace std; // for cout without std::

namespace PSQt {

//--------------------------

GUILogger::GUILogger( QWidget *parent, const bool& showbuts, const bool& showframe)
  : Frame(parent) //  , Logger()
  , m_showbuts(showbuts)
  , m_showframe(showframe)
{
  m_txt_edi  = new QTextEdit();
  m_but_save = new QPushButton("Save");
  m_combo    = new QComboBox(this);

  m_list << strLevel(DEBUG).c_str()
         << strLevel(INFO).c_str()
         << strLevel(WARNING).c_str()
         << strLevel(ERROR).c_str()
         << strLevel(CRITICAL).c_str();

  m_combo -> addItems(m_list);
  m_combo -> setCurrentIndex(1); 

  m_cbox = new QHBoxLayout;
  m_cbox -> addStretch(1);
  m_cbox -> addWidget(m_combo);
  m_cbox -> addWidget(m_but_save);

  m_vbox = new QVBoxLayout;
  m_vbox -> addWidget(m_txt_edi);
  m_vbox -> addLayout(m_cbox);

  this -> setLayout(m_vbox);

  this -> showTips();
  this -> setStyle();
  this -> addStartRecords();

  connect(Logger::getLogger(), SIGNAL(signal_new_record(Record&)), this, SLOT(addNewRecord(Record&)));
  connect(m_combo, SIGNAL(currentIndexChanged(int)), this, SLOT(onCombo(int)) );
  connect(m_but_save, SIGNAL(clicked()), this, SLOT(onSave()) );
}

//--------------------------
void
GUILogger::showTips() 
{
  m_combo->setToolTip("Select the level of messages to show");
  m_txt_edi->setToolTip("Log messages");
  m_but_save->setToolTip("Save log in file");
}

//--------------------------
void
GUILogger::setStyle() 
{
  setGeometry(100, 100, 500, 300);

  m_but_save->setVisible(m_showbuts);
  m_combo   ->setVisible(m_showbuts);

  Frame::setBoarderVisible(m_showframe);
  if (! m_showframe) this -> setContentsMargins(-9,-9,-9,-9);

  //this -> setWindowTitle(tr("GUILogger"));
  //this -> setMinimumWidth(700);
  //this -> setFixedWidth(220);
  //this -> setFixedHeight(50);
  //this -> setFixedHeight( (m_show_frame)? 50 : 34);

  //this -> move(300,50);
}

//--------------------------
void 
GUILogger::resizeEvent(QResizeEvent *event)
{
  stringstream ss; ss << "w:" << event->size().width() << " h:" <<  event->size().height();
  setWindowTitle(ss.str().c_str());
}

//--------------------------
void
GUILogger::moveEvent(QMoveEvent *event)
{
  stringstream ss; ss << "x:" << event->pos().x() << " y:" << event->pos().y();
  setWindowTitle(ss.str().c_str());
}

//--------------------------
void 
GUILogger::closeEvent(QCloseEvent *event)
{
  QWidget::closeEvent(event);
  stringstream ss; ss << "closeEvent(...): type = " << event -> type();
  MsgInLog(_name_(), INFO, ss.str());
}

//--------------------------
void  
GUILogger::onCombo(int i)
{
  std::string str_level = m_list.at(i).toStdString();
  MsgInLog(_name_(), INFO, "Selected level: " + str_level);

  LEVEL level = levelFromString(str_level);
  std::string recs = Logger::getLogger()->strRecordsForLevel(level);
  m_txt_edi->clear();
  m_txt_edi->append(recs.c_str());

  SetMsgLevel(level);

  scrollDown();
}

//--------------------------
void  
GUILogger::onSave()
{
  MsgInLog(_name_(), INFO, "\"Save\" button is clicked");
  SaveLog();
}

//--------------------------
void 
GUILogger::scrollDown()
{
  m_txt_edi->moveCursor(QTextCursor::End);
  m_txt_edi->repaint();
}

//--------------------------
void 
GUILogger::addNewRecord(Record& rec)
{
  //std::cout << rec.strRecordTotal() << '\n';  
  //m_txt_edi->append(rec.strRecordTotal().c_str());
  m_txt_edi->append(rec.strRecord().c_str());

  scrollDown();
}


//--------------------------
void 
GUILogger::addStartRecords()
{
  std::string start_recs = Logger::getLogger()->strRecordsForLevel(Logger::getLogger()->getLevel());
  m_txt_edi->append(start_recs.c_str());

  //cout << _name_() << "addStartRecords():\n" << start_recs << '\n';

  scrollDown();
}

//--------------------------
//--------------------------
//--------------------------
//--------------------------

} // namespace PSQt

//--------------------------

File: CalibManager/tags/V00-00-92/src/GUILogger.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id$
#
# Description:
#  Module GUILogger
#
#------------------------------------------------------------------------

"""GUILogger - GUI for logger

This software was developed for the SIT project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@see RelatedModule

@version $Id$

@author Mikhail S. Dubrovin
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision$"
# $Source$

#----------------------
#  Import of modules --
#----------------------

import os
import sys

from ConfigParametersForApp import cp
from Logger import logger
from PyQt4 import QtGui, QtCore

#-----------------------------

class GUILogger ( QtGui.QWidget ) :
    """GUI for Logger"""

    name = 'GUILogger'

    def __init__ ( self, parent=None, show_buttons=True ) :

        QtGui.QWidget.__init__(self, parent)

        self.show_buttons = show_buttons
        
        self.setGeometry(200, 400, 500, 400)
        self.setWindowTitle('GUI Logger')
        #self.setWindowIcon(cp.icon_logger)

        self.setFrame()

        self.box_txt    = QtGui.QTextEdit()
 
        #self.tit_title  = QtGui.QLabel('Logger')
        self.tit_status = QtGui.QLabel('Status:')
        self.tit_level  = QtGui.QLabel('Verbosity level:')
        self.but_close  = QtGui.QPushButton('&Close') 
        self.but_save   = QtGui.QPushButton('&Save log-file') 

        self.list_of_levels = logger.getListOfLevels()
        self.box_level      = QtGui.QComboBox( self ) 
        self.box_level.addItems(self.list_of_levels)
        self.box_level.setCurrentIndex( self.list_of_levels.index(cp.log_level.value()) )
        
        self.hboxM = QtGui.QHBoxLayout()
        self.hboxM.addWidget( self.box_txt )

        self.hboxB = QtGui.QHBoxLayout()
        self.hboxB.addWidget(self.tit_status)
        self.hboxB.addStretch(4)     
        self.hboxB.addWidget(self.tit_level)
        self.hboxB.addWidget(self.box_level)
        self.hboxB.addStretch(1)     
        self.hboxB.addWidget(self.but_save)
        self.hboxB.addWidget(self.but_close)

        self.vbox  = QtGui.QVBoxLayout()
        #self.vbox.addWidget(self.tit_title)
        self.vbox.addLayout(self.hboxM)
        self.vbox.addLayout(self.hboxB)
        self.setLayout(self.vbox)
        
        self.connect( self.but_close, QtCore.SIGNAL('clicked()'), self.onClose )
        self.connect( self.but_save,  QtCore.SIGNAL('clicked()'), self.onSave  )
        self.connect( self.box_level, QtCore.SIGNAL('currentIndexChanged(int)'), self.onBox  )
 
        self.startGUILog()

        self.showToolTips()
        self.setStyle()

        self.guilogger = self

    #-------------------
    #  Public methods --
    #-------------------

    def showToolTips(self):
        #self           .setToolTip('This GUI is for browsing log messages')
        self.box_txt    .setToolTip('Window for log messages')
        self.but_close  .setToolTip('Close this window')
        self.but_save   .setToolTip('Save current content of the GUI Logger\nin work directory file: '+os.path.basename(self.fname_log))
        self.tit_status .setToolTip('The file name, where this log \nwill be saved at the end of session')
        self.box_level  .setToolTip('Click on this button and \nselect the level of messages \nwhich will be displayed')


    def setFrame(self):
        self.frame = QtGui.QFrame(self)
        self.frame.setFrameStyle( QtGui.QFrame.Box | QtGui.QFrame.Sunken ) #Box, Panel | Sunken, Raised 
        self.frame.setLineWidth(0)
        self.frame.setMidLineWidth(1)
        self.frame.setGeometry(self.rect())
        #self.frame.setVisible(False)


    def setStyle(self):
        self.           setStyleSheet (cp.styleBkgd)
        #self.tit_title.setStyleSheet (cp.styleTitleBold)
        self.tit_status.setStyleSheet (cp.styleTitle)
        self.tit_level .setStyleSheet (cp.styleTitle)
        self.but_close .setStyleSheet (cp.styleButton)
        self.but_save  .setStyleSheet (cp.styleButton) 
        self.box_level .setStyleSheet (cp.styleButton) 
        self.box_txt   .setReadOnly(True)
        self.box_txt   .setStyleSheet (cp.styleWhiteFixed) 
        #self.box_txt   .ensureCursorVisible()
        #self.tit_title.setAlignment(QtCore.Qt.AlignCenter)
        #self.titTitle.setBold()

        self.tit_status.setVisible(self.show_buttons)
        self.tit_level .setVisible(self.show_buttons)
        self.box_level .setVisible(self.show_buttons)
        self.but_save  .setVisible(self.show_buttons)
        self.but_close .setVisible(self.show_buttons)

        if not self.show_buttons : self.setContentsMargins (QtCore.QMargins(-9,-9,-9,-9))
        self.setMinimumHeight(50)
        self.setMinimumSize(300,50)
        #self.setBaseSize(500,200)


    def setParent(self,parent) :
        self.parent = parent


    def resizeEvent(self, e):
        #logger.debug('resizeEvent', self.name) 
        self.frame.setGeometry(self.rect())


    def moveEvent(self, e):
        #logger.debug('moveEvent', self.name) 
        #cp.posGUIMain = (self.pos().x(),self.pos().y())
        pass


    def closeEvent(self, event):
        logger.info('closeEvent', self.name)
        #self.saveLogTotalInFile() # It will be saved at closing of GUIMain

        #try    : del cp.guilogger # GUILogger
        #except : pass

        #if cp.guilogger is not None :
        #    del cp.guilogger
        #    cp.guilogger = None


    def onClose(self):
        logger.debug('onClose', self.name)
        self.close()


    def onSave(self):
        logger.debug('onSave:', self.name)
        self.saveLogInFile()


    def onBox(self):
        level_selected = self.box_level.currentText()
        cp.log_level.setValue( level_selected ) 
        logger.info('onBox - selected ' + self.tit_level.text() + ' ' + cp.log_level.value(), self.name)
        logger.setLevel(cp.log_level.value())
        self.box_txt.setText( logger.getLogContent() )


    def saveLogInFile(self):
        logger.info('saveLogInFile ' + self.fname_log, self.name)
        path = str( QtGui.QFileDialog.getSaveFileName(self,
                                                      caption   = 'Select the file to save log',
                                                      directory = self.fname_log,
                                                      filter    = '*.txt'
                                                      ) )
        if path == '' :
            logger.debug('Saving is cancelled.', self.name)
            return 
        logger.info('Output file: ' + path, self.name)
        logger.saveLogInFile(path)
        self.fname_log = path
        cp.log_file.setValue(path)


    def saveLogTotalInFile(self):
        logger.info('saveLogTotalInFile' + self.fname_log_total, self.name)
        logger.saveLogTotalInFile(self.fname_log_total)


    def getConfirmation(self):
        """Pop-up box for confirmation"""
        msg = QtGui.QMessageBox(self, windowTitle='Confirm closing!',
            text='You are about to close GUI Logger...\nIf the log-file is not saved it will be lost.',
            standardButtons=QtGui.QMessageBox.Save | QtGui.QMessageBox.Discard | QtGui.QMessageBox.Cancel)
        msg.setDefaultButton(msg.Save)

        clicked = msg.exec_()

        if   clicked == QtGui.QMessageBox.Save :
            logger.info('Saving is requested', self.name)
        elif clicked == QtGui.QMessageBox.Discard :
            logger.info('Discard is requested', self.name)
        else :
            logger.info('Cancel is requested', self.name)
        return clicked


    def onShow(self):
        logger.info('onShow - is not implemented yet...', self.name)


    def startGUILog(self) :
        self.fname_log       = cp.log_file.value()
        #self.fname_log_total = cp.log_file_total.value()
        self.setStatus(0, 'Log-file: ' + os.path.basename(self.fname_log))

        logger.setLevel(cp.log_level.value())
        self.box_txt.setText(logger.getLogContent())
        
        logger.setGUILogger(self)
        logger.debug('GUILogger is open', self.name)
        self.box_txt.moveCursor(QtGui.QTextCursor.End)


    def appendGUILog(self, msg='...'):
        self.box_txt.append(msg)
        self.scrollDown()


    def scrollDown(self):
        #print 'scrollDown'
        #scrol_bar_v = self.box_txt.verticalScrollBar() # QScrollBar
        #scrol_bar_v.setValue(scrol_bar_v.maximum()) 
        self.box_txt.moveCursor(QtGui.QTextCursor.End)
        self.box_txt.repaint()
        #self.raise_()
        #self.box_txt.update()

        
    def setStatus(self, status_index=0, msg=''):
        list_of_states = ['Good','Warning','Alarm']
        if status_index == 0 : self.tit_status.setStyleSheet(cp.styleStatusGood)
        if status_index == 1 : self.tit_status.setStyleSheet(cp.styleStatusWarning)
        if status_index == 2 : self.tit_status.setStyleSheet(cp.styleStatusAlarm)

        #self.tit_status.setText('Status: ' + list_of_states[status_index] + msg)
        self.tit_status.setText(msg)

#-----------------------------

def test_GUILogger() :
    app = QtGui.QApplication(sys.argv)
    widget = GUILogger ()
    widget.show()
    app.exec_()

#-----------------------------

if __name__ == "__main__" :

    test_GUILogger()
    sys.exit (0)

#-----------------------------
